modules/caddytls/internalissuer.go

```diff
@@ -0,0 +1,199 @@
+func (li InternalIssuer) Issue(ctx context.Context, csr *x509.CertificateRequest) (*certmagic.IssuedCertificate, error) {
+	// prepare the signing authority
+	// TODO: eliminate placeholders / needless values
+	cfg := &authority.Config{
+		Address:          "placeholder_Address:1",
+		Root:             []string{"placeholder_Root"},
+		IntermediateCert: "placeholder_IntermediateCert",
+		IntermediateKey:  "placeholder_IntermediateKey",
+		DNSNames:         []string{"placeholder_DNSNames"},
+		AuthorityConfig: &authority.AuthConfig{
+			Provisioners: provisioner.List{},
+		},
+	}
+
+	// get the root certificate and the issuer cert+key
+	rootCert := li.ca.RootCertificate()
+	var issuerCert *x509.Certificate
+	var issuerKey interface{}
+	if li.SignWithRoot {
+		issuerCert = rootCert
+		var err error
+		issuerKey, err = li.ca.RootKey()
+		if err != nil {
+			return nil, fmt.Errorf("loading signing key: %v", err)
+		}
+	} else {
+		issuerCert = li.ca.IntermediateCertificate()
+		issuerKey = li.ca.IntermediateKey()
+	}
+
+	auth, err := authority.New(cfg,
+		authority.WithX509Signer(issuerCert, issuerKey.(crypto.Signer)),
+		authority.WithX509RootCerts(rootCert),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("initializing certificate authority: %v", err)
+	}
+
+	// ensure issued certificate does not expire later than its issuer
+	lifetime := time.Duration(li.Lifetime)
+	if time.Now().Add(lifetime).After(issuerCert.NotAfter) {
+		// TODO: log this
+		lifetime = issuerCert.NotAfter.Sub(time.Now())
+	}
+
+	certChain, err := auth.Sign(csr, provisioner.Options{},
+		profileDefaultDuration(li.Lifetime),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	var buf bytes.Buffer
+	for _, cert := range certChain {
+		err := pem.Encode(&buf, &pem.Block{Type: "CERTIFICATE", Bytes: cert.Raw})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &certmagic.IssuedCertificate{
+		Certificate: buf.Bytes(),
+	}, nil
+}
+
+// TODO: borrowing from https://github.com/smallstep/certificates/blob/806abb6232a5691198b891d76b9898ea7f269da0/authority/provisioner/sign_options.go#L191-L211
+// as per https://github.com/smallstep/certificates/issues/198.
+// profileDefaultDuration is a wrapper against x509util.WithOption to conform
+// the SignOption interface.
+type profileDefaultDuration time.Duration
+
+// TODO: is there a better way to set cert lifetimes than copying from the smallstep libs?
+func (d profileDefaultDuration) Option(so provisioner.Options) x509util.WithOption {
+	var backdate time.Duration
+	notBefore := so.NotBefore.Time()
+	if notBefore.IsZero() {
+		notBefore = time.Now().Truncate(time.Second)
+		backdate = -1 * so.Backdate
+	}
+	notAfter := so.NotAfter.RelativeTime(notBefore)
+	return func(p x509util.Profile) error {
+		fn := x509util.WithNotBeforeAfterDuration(notBefore, notAfter, time.Duration(d))
+		if err := fn(p); err != nil {
+			return err
+		}
+		crt := p.Subject()
+		crt.NotBefore = crt.NotBefore.Add(backdate)
+		return nil
+	}
+}
```

Comment:
@mmalone @maraino It would be nice if there was an easier way to simply specify the certificate lifetime. I just feel kinda not-awesome copying unexported code out of your repo, especially for keeping up with changes. Open to discussing ideas here!